File: client/ts/src/rpc/listTree.ts
import { buildTextRpc } from './TextRpc';

interface ListedTreeNode {
  type: 'node';
  locator: NodeLocator;
  name?: string;
  children: (ListedTreeNode[]) | { type: 'placeholder', num: number };
}
export { ListedTreeNode }
export default buildTextRpc<{
  attr: { name: 'meta:listTreeUpwards' | 'meta:listTreeDownwards'; };
  locator: NodeLocator;
}, {
  body?: RpcBodyLine[];
  locator: NodeLocator;
  nodes: ListedTreeNode;
}>();

File: protocol/src/protocolgen/spec/ListTree.java
package protocolgen.spec;

@SuppressWarnings("unused")
public class ListTree extends Rpc {

	@Override
	public Streamable getRequestType() {
		return new Streamable() {
			public final Object type = oneOf("ListTreeUpwards", "ListTreeDownwards");
			public final Object locator = NodeLocator.class;
			public final Object src = ParsingRequestData.class;
		};
	}

	@Override
	public Streamable getResponseType() {
		return new Streamable() {
			public final Object body = arr(RpcBodyLine.class);
			public final Object locator = opt(NodeLocator.class);
			public final Object node = opt(ListedTreeNode.class);
		};
	}
}
